### PR TITLE
Secret Gateways: Config loaded Away Missions + Anti-observing Z level traits

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request

Adds secret gateways, these away missions are map files loaded from a config folder, instead of being files in the repo.
The chance of an away mission being a secret one is config dependant.
Secret away missions have a secret z level trait applied to them, making them not visible by ghosts (other than admin ghosts) other than looking through another player's eyes (you dont get xray like you normally do) with the Observe verb like in the image below.
image
Until there are maps added to the folder and someone sets the config up, no secret away missions will show up on the servers.
Adds a map report verb for viewing map reports
image
If the map has errors in it, it doesn't load and it creates a map report
image

## Why It's Good For The Game

content that cant be code dived seems fun at least to try out.

## Changelog

:cl:
add: Secret Gateways can now be a thing enabled by config. They are not on the repository, instead, loaded from a config folder.
add: Adds the secret and anti-phasing z level traits used by Secret Gateways, secret prevents observers from looking at the z level in any way other than observing through another player's eyes
admin: admin loaded away missions can be secret as well
/:cl:

### CODE REVIEW

File: code/modules/mob/mob_movement.dm
Comment (Watermelon914) [inline code/modules/mob/mob_movement.dm:557]: on second thought, have this be as some sort of zlevel trait instead.

File: code/modules/mob/dead/observer/observer.dm
Comment (Ghommie) [inline code/modules/mob/dead/observer/observer.dm:901]: This signal ought to be unregistered in `reset_perspective()`

File: code/modules/vehicles/mecha/_mecha.dm
Comment (Watermelon914) [inline code/modules/vehicles/mecha/_mecha.dm:279]: have this be as some different zlevel trait instead, there may be times when it is useful to prevent phasing for an entire zlevel, don't know but I don't think it needs to be exclusive to the secret ztrait.

Comment (Watermelon914) [reply code/modules/vehicles/mecha/_mecha.dm:279]: There may also be times when you might want an away mission to be jauntable.
Maybe apply this to non-secret away missions too.

File: code/modules/mob/living/death.dm
Comment (Ghommie) [inline code/modules/mob/living/death.dm:77]: This is bad way to stop players from peeking behind the walls while dead. It'll effectively impair revival of the mob considering it requires the player to be in control of the mob and not a ghost.
That said, it's better to just do some changes to `update_sight()` so it won't grant (SEE_MOBS|SEE_OBJS|SEE_TURFS) to dead mobs while they're in a sekrit zlevel.

File: code/modules/mob/dead/observer/observer.dm
Comment (Ghommie) [inline code/modules/mob/dead/observer/observer.dm:368]: See above.

File: code/modules/mob/living/death.dm
Comment (Fikou) [reply code/modules/mob/living/death.dm:77]: not true defibs no longer require ghost to be in body and you can return to it when your body comes back from death anyway

File: code/modules/mob/dead/observer/observer.dm
Comment (Fikou) [reply code/modules/mob/dead/observer/observer.dm:901]: oh right i forgot thanks

File: code/modules/mob/living/death.dm
Comment (Ghommie) [reply code/modules/mob/living/death.dm:77]: > 
> 
> not true defibs no longer require ghost to be in body and you can return to it when your body comes back from death anyway

Since when?

**Comment by wesoda25** [comment]
Athath nerf

**Comment by ATH1909** [comment]
> Athath nerf

:walter:

File: code/modules/mob/living/death.dm
Comment (Fikou) [reply code/modules/mob/living/death.dm:77]: idk since like a few months but sure ill look into it more

**Comment by MMMiracles** [comment]
is there any talk about how these will be submitted/maintained since the idea is that they aren't easily seen on the repo? i think this is a really good step towards making more long-term away missions and knowing they won't be immediately mapped out within a week gives me a lot of hope for their revival

**Comment by Fikou** [comment]
> is there any talk about how these will be submitted/maintained since the idea is that they aren't easily seen on the repo? i think this is a really good step towards making more long-term away missions and knowing they won't be immediately mapped out within a week gives me a lot of hope for their revival

this depends on the headmins but im hopeful we can get a good system going for it, would be good to start a policy discussion thread on it when the forums go back up maybe

**Comment by Mothblocks** [comment]
This is pretty spooky since changing things like paths or var-edits is going to be completely unnoticed, especially since gateways are only loaded at specific times.

It's going to effectively be mandatory now if we want these up to date that the path change tool is always updated even for small changes, though even that won't make var-edit changes possible. Any big refactor like that disposals one (I think?) will just break these until someone A. notices and B. has the power to fix it. (Also, I don't even think the path change tool, works.)

It's good to see the map report tools and that it's blocked from loading if it doesn't align, I'm just worried about the future. I'm not blocking this.

File: code/__DEFINES/maps.dm
Comment (Mothblocks) [inline code/__DEFINES/maps.dm:47]:
```suggestion
/// boolean - does this z prevent ghosts from observing it
```

File: code/__DEFINES/maps.dm
Comment (Mothblocks) [inline code/__DEFINES/maps.dm:50]:
```suggestion
/// boolean - does this z prevent phasing
```

File: code/__DEFINES/maps.dm
Comment (Mothblocks) [inline code/__DEFINES/maps.dm:86]: Document both these.

File: code/modules/awaymissions/zlevel.dm
Comment (Mothblocks) [inline code/modules/awaymissions/zlevel.dm:15]: Can we try to load a different, potentially non-config gateway instead?

File: code/modules/awaymissions/zlevel.dm
Comment (Mothblocks) [inline code/modules/awaymissions/zlevel.dm:74]:
```suggestion
		if(!findtext(map_file, ".dmm"))
```

I don't think there's any reason to check specifically markdown files.

File: code/modules/mapping/verify.dm
Comment (Mothblocks) [inline code/modules/mapping/verify.dm:1]: Document.

File: code/modules/mob/dead/observer/observer.dm
Comment (Mothblocks) [inline code/modules/mob/dead/observer/observer.dm:911]: SIGNAL_HANDLER

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by MrDoomBringer** [comment]
> This is pretty spooky since changing things like paths or var-edits is going to be completely unnoticed, especially since gateways are only loaded at specific times.

Would another layer of security be to host these maps on a private repo?

**Comment by Fikou** [comment]
> > This is pretty spooky since changing things like paths or var-edits is going to be completely unnoticed, especially since gateways are only loaded at specific times.
> 
> Would hosting these maps on a private downstream repo help us catch broken maps faster?

broken maps should be caught pretty fast with the game reporting them and the reports being viewable, the issue is getting an "input and output" from server files, so we have people that care about maintaining the maps can do it

**Comment by drIcepop** [comment]
That first picture looks like Biodome, telesci clause in TG?!?!?!

**Comment by Watermelon914** [comment]
> This is pretty spooky since changing things like paths or var-edits is going to be completely unnoticed, especially since gateways are only loaded at specific times.
> 
> It's going to effectively be mandatory now if we want these up to date that the path change tool is always updated even for small changes, though even that won't make var-edit changes possible. Any big refactor like that disposals one (I think?) will just break these until someone A. notices and B. has the power to fix it. (Also, I don't even think the path change tool, works.)
> 
> It's good to see the map report tools and that it's blocked from loading if it doesn't align, I'm just worried about the future. I'm not blocking this.

It's up to the headadmins how they want to run this but I was thinking they rotate maps every few weeks or so instead of having a map in forever.
If a map deserves to be in the configs for more than 2-3 months, then it should probably just be PR'd to the repo as an away mission.

**Comment by Fikou** [comment]
> That first picture looks like Biodome, telesci clause in TG?!?!?!

so true bestie

**Comment by Timberpoes** [comment]
Ultimately it will be up to the headmins and MSO to set up the infrastructure and team for uploading and maintaining these private gateway maps.

The private gateway map team will be required to update and maintain these private gateway maps independent of the codebase.

In addition with the new z-level trait that hides it from observers - I often stand alone when I say I'd like to limit observer information so they are not omnipotent. People from up and down the admin team and maintainer team, as well as MSO, have expressed discontent at attempts taken to restrict or remove observer's information. I feel that should also be discussed before this is merged.

**Comment by Fikou** [comment]
well timber we cant really do that until someone that dislikes restricting observers shows up here!

**Comment by Mothblocks** [comment]
We *already* limit observer information *especially* for admin stuff. Admin circuits don't let you look inside them because of loser powergamers.

**Comment by adamkad1** [comment]
What is this, goonstation?

**Comment by Mothblocks** [comment]
no sir you are on the tgstation repo. https://github.com/goonstation/goonstation <-- this is goonstation

**Pull request merged**
